src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -0,0 +1,284 @@
+            boolConversion = Conversion.NoConversion;
+            boolOperator = default;
+            return;
+        }
+
+        private TupleBinaryOperatorInfo BindTupleDynamicBinaryOperatorSingleInfo(BinaryExpressionSyntax node, BinaryOperatorKind kind,
+            BoundExpression left, BoundExpression right, DiagnosticBag diagnostics)
+        {
+            // This method binds binary == and != operators where one or both of the operands are dynamic.
+            Debug.Assert((object)left.Type != null && left.Type.IsDynamic() || (object)right.Type != null && right.Type.IsDynamic());
+
+            bool hasError = false;
+            if (!IsLegalDynamicOperand(left) || !IsLegalDynamicOperand(right))
+            {
+                // Operator '{0}' cannot be applied to operands of type '{1}' and '{2}'
+                Error(diagnostics, ErrorCode.ERR_BadBinaryOps, node, node.OperatorToken.Text, left.Display, right.Display);
+                hasError = true;
+            }
+
+            BinaryOperatorKind elementOperatorKind = hasError ? kind : kind.WithType(BinaryOperatorKind.Dynamic);
+            TypeSymbol dynamicType = Compilation.DynamicType;
+
+            // We'll want to dynamically invoke operators op_true (/op_false) for equality (/inequality) comparison, but we don't need
+            // to prepare either a conversion or a truth operator. Those can just be synthesized during lowering.
+            return new TupleBinaryOperatorInfo.Single(dynamicType, dynamicType, elementOperatorKind,
+                methodSymbolOpt: null, boolConversion: Conversion.Identity, boolOperator: default);
+        }
+
+        private TupleBinaryOperatorInfo BindTupleBinaryOperatorNestedInfo(BinaryExpressionSyntax node, BinaryOperatorKind kind,
+            ImmutableArray<BoundExpression> left, ImmutableArray<BoundExpression> right,
+            bool isNullable, DiagnosticBag diagnostics)
+        {
+            int length = left.Length;
+            Debug.Assert(length == right.Length);
+
+            var operatorsBuilder = ArrayBuilder<TupleBinaryOperatorInfo>.GetInstance(length);
+
+            for (int i = 0; i < length; i++)
+            {
+                operatorsBuilder.Add(BindTupleBinaryOperatorInfo(node, kind, left[i], right[i], diagnostics));
+            }
+
+            var compilation = this.Compilation;
+            var operators = operatorsBuilder.ToImmutableAndFree();
+
+            TypeSymbol leftTupleType = MakeConvertedType(operators.SelectAsArray(o => o.LeftConvertedType), node.Left, left, isNullable, compilation, diagnostics, isRight: false);
+            TypeSymbol rightTupleType = MakeConvertedType(operators.SelectAsArray(o => o.RightConvertedType), node.Right, right, isNullable, compilation, diagnostics, isRight: true);
+
+            return new TupleBinaryOperatorInfo.Multiple(operators, leftTupleType, rightTupleType);
+        }
+
+        private static int? GetTupleCardinality(BoundExpression expr)
```

Comment:
>int? [](start = 23, length = 4)

Perhaps we can return 0 instead of null and avoid the overhead of dealing with nullable?

Reply:
Yes, I've already made that change in follow-up PR (see last commit of https://github.com/dotnet/roslyn/pull/25076). I have it returning `-1`.